afl-fuzz.c

```diff
@@ -1330,31 +1461,106 @@ static void cull_queue(void) {
 
   q = queue;
 
-  while (q) {
-    q->favored = 0;
-    q = q->next;
-  }
+  // use AFL's original mechanism to assign favorites
+  if (disable_random_favorites) {
+
+    while (q) {
+      if (disable_afl_default_favorites)
+        q->favored = 1;
+      else
+        q->favored = 0;
+      q = q->next;
+    }
 
-  /* Let's see if anything in the bitmap isn't captured in temp_v.
-     If yes, and if it has a top_rated[] contender, let's use it. */
+    /* Let's see if anything in the bitmap isn't captured in temp_v.
+      If yes, and if it has a top_rated[] contender, let's use it. */
 
-  for (i = 0; i < MAP_SIZE; i++)
-    if (top_rated[i] && (temp_v[i >> 3] & (1 << (i & 7)))) {
+    for (i = 0; i < MAP_SIZE; i++)
+      if (top_rated[i] && (temp_v[i >> 3] & (1 << (i & 7)))) {
+
+        u32 j = MAP_SIZE >> 3;
+
+        /* Remove all bits belonging to the current entry from temp_v. */
+
+        while (j--) 
+          if (top_rated[i]->trace_mini[j])
+            temp_v[j] &= ~top_rated[i]->trace_mini[j];
 
-      u32 j = MAP_SIZE >> 3;
+        top_rated[i]->favored = 1;
+        queued_favored++;
 
-      /* Remove all bits belonging to the current entry from temp_v. */
+        if (!top_rated[i]->was_fuzzed) pending_favored++;
 
-      while (j--) 
-        if (top_rated[i]->trace_mini[j])
-          temp_v[j] &= ~top_rated[i]->trace_mini[j];
+      }
+      
+  } else {
+    // otherwise, randomly assign favorites
+    int r;
+    int rid;
+    double weight;
+
+    while (q) {
+      weight = 1.0;
+      if (!enable_uniformly_random_favorites) {
+        // enable_boost_inputs
+        double base_weight_fac_boost_inputs = 1.0;
+        double max_weight_fac_incr = 7.0;
+        double scale_fac_boost_inputs = 0.001;
+        double num_selections = (double)q->num_fuzzed;
+        weight *= base_weight_fac_boost_inputs + max_weight_fac_incr / (scale_fac_boost_inputs * num_selections + 1.0);
+
+        // enable_boost_fast_seqs
+        double base_weight_fac_boost_fast = 8.0;
+        double max_weight_fac_decr = 7.0;
+        double scale_fac_boost_fast = 0.005;
```

Comment:
@jiradeto Can we try changing this value?

Possible alternatives to try:
- 0.0025
- 0.001
- 0.0005

Let's use these configs:
1. afl (vanilla AFL)
2. afl_wrs_rf: !DISABLE_WRS && !DISABLE_RF && !ENABLE_UF && !DISABLE_FAVS && DISABLE_RP && scale_fac_boost_fast=0.005
3. afl_wrs_rf_25: !DISABLE_WRS && !DISABLE_RF && !ENABLE_UF && !DISABLE_FAVS && DISABLE_RP && scale_fac_boost_fast=0.0025
4. afl_wrs_rf_1: !DISABLE_WRS && !DISABLE_RF && !ENABLE_UF && !DISABLE_FAVS && DISABLE_RP && scale_fac_boost_fast=0.001
5. afl_wrs_rf_05: !DISABLE_WRS && !DISABLE_RF && !ENABLE_UF && !DISABLE_FAVS && DISABLE_RP && scale_fac_boost_fast=0.0005